bob-access/src/authenticator/basic.rs

```diff
@@ -1,62 +1,141 @@
+    resolve_sleep_dur_ms: u64,
 }
 
 impl<Storage: UsersStorage> Basic<Storage> {
-    pub fn new(users_storage: Storage) -> Self {
+    pub fn new(users_storage: Storage, resolve_sleep_dur_ms: u64) -> Self {
         Self {
             users_storage,
-            nodes: HashMap::new(),
+            nodes: Arc::new(RwLock::new(HashMap::new())),
+            resolve_sleep_dur_ms,
         }
     }
 
+    fn node_creds_ok(creds: &HashMap<String, Credentials>) -> bool {
+        creds.values()
+            .all(|cred|
+                (cred.ip().is_some() || cred.hostname().is_some()) &&
+                cred.kind().map(|k| k.is_internode()) == Some(true))
+    }
+
     pub fn set_nodes_credentials(
         &mut self,
-        nodes: HashMap<IpAddr, Vec<Credentials>>,
+        mut nodes: HashMap<String, Credentials>,
     ) -> Result<(), Error> {
-        if nodes
-            .values()
-            .all(|creds|
-                creds
-                    .iter()
-                    .all(|cred|
-                        cred.ip().is_some() &&
-                        cred.kind().map(|k| k.is_internode()) == Some(true)))
-        {
-            self.nodes = nodes;
+        if Self::node_creds_ok(&nodes) {
+            let mut resolved = HashMap::new();
+            for (nodename, cred) in nodes.drain() {
+                if cred.ip().is_none() && cred.hostname().is_some() {
+                    self.spawn_resolver(cred);
+                } else {
+                    resolved.insert(nodename, cred);
+                }
+            }
+            let mut nodes_creds = self.nodes.write().expect("nodes credentials lock");
+            nodes_creds.extend(resolved.drain());
             Ok(())
         } else {
             let message = "nodes credentials missing ip or node name";
             Err(Error::CredentialsNotProvided(message.to_string()))
         }
     }
 
-    fn check_node_request(&self, node_name: &String, ip: Option<IpAddr>) -> Option<bool> {
-        if self.nodes.is_empty() {
-            warn!("nodes credentials not set");
+    fn spawn_resolver(&self, cred: Credentials) {
+        tokio::spawn(Self::resolve_worker(self.nodes.clone(), cred, self.resolve_sleep_dur_ms));
+    }
+
+    async fn resolve_worker(nodes: Arc<RwLock<NodesCredentials>>, mut cred: Credentials, sleep_dur_ms: u64) {
+        let hostname = cred.hostname().as_ref().expect("resolve worker without hostname");
+        let mut addr: Option<Vec<SocketAddr>> = None;
+        for _ in 0..60 {
+            addr = match lookup_host(hostname).await {
+                Ok(address) => Some(address.collect()),
+                _ => None
+            };
+            if let Some(addr) = addr.as_ref() {
+                if addr.len() > 0 {
+                    break;
+                }
+            }
+
+            tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+
+        while addr.is_none() || (addr.is_some() && addr.as_ref().unwrap().len() == 0) {
```

Comment:
You have 2 loops that do the same thing with different sleep interval. You should keep only one.
Sleep duration update strategy can be improved: you can start with 100ms duration and on every iteration multiply it by 2, but check that it will not become larger than `sleep_dur_ms` (`sleep_dur_ms.min(cur_sleep_dur * 2)`)